packages/flutter_tools/templates/ffi_plugin/android.tmpl/build.gradle.tmpl

```diff
@@ -0,0 +1,41 @@
+group '{{androidIdentifier}}'
+version '1.0'
+
+buildscript {
+    repositories {
+        google()
+        mavenCentral()
+    }
+
+    dependencies {
+        classpath 'com.android.tools.build:gradle:4.1.0'
+    }
+}
+
+rootProject.allprojects {
+    repositories {
+        google()
+        mavenCentral()
+    }
+}
+
+apply plugin: 'com.android.library'
+
+android {
+    compileSdkVersion 30
+
+    externalNativeBuild {
+        cmake {
+            path "../src/CMakeLists.txt"
```

Comment:
We need a native build.

Since we are already using CMake for Flutter Desktop on Linux/Windows, CMake is a logical choice.

CMake comes as part of the Android NDK. I'm not sure about the SDK. Without the NDK people cannot build native apps/components for Android.

Maybe we should add the NDK to Flutter doctor.

Reply:
> Ideally, the check would only fire if needed (e.g. someone is doing an FFI based plugin)

The problem is that "doing" means "using, maybe even transitively" not "developing", so there's really no way for people to know in advance.

We will likely need to do a build-time check when we see that there's an Android FFI plugin so that we can do a clear error message, since I'm assuming otherwise it's going to be super confusing (e.g., something like a gradle message about not finding CMake, which someone building an app that happens to use an FFI plugin isn't going to understand).

Reply:
> I think we should have an NDK check in Flutter doctor before we land this.

The AGP (Android Gradle Plugin) automatically downloads the NDK when it needs to. We are already pinning AGP 4.1.0 in all our templates (`com.android.tools.build:gradle:4.1.0`). AGP 4.1 [has pinned the NDK version](https://developer.android.com/studio/projects/install-ndk#default-ndk-per-agp) to `21.1.6352462`.

After running `flutter doctor` being happy with the Android toolchain and `flutter doctor --android-licenses` without an NDK installed, running an app using an FFI plugin will automatically download the NDK:

```
Using hardware rendering with device sdk gphone x86. If you notice graphics artifacts, consider enabling software rendering with "--enable-software-rendering".
Launching lib/main.dart on sdk gphone x86 in debug mode...
Checking the license for package NDK (Side by side) 21.1.6352462 in /Users/dacoharkes/Library/Android/sdk/licenses
License for package NDK (Side by side) 21.1.6352462 accepted.
Preparing "Install NDK (Side by side) 21.1.6352462 (revision: 21.1.6352462)".
"Install NDK (Side by side) 21.1.6352462 (revision: 21.1.6352462)" ready.
Installing NDK (Side by side) 21.1.6352462 in /Users/dacoharkes/Library/Android/sdk/ndk/21.1.6352462
"Install NDK (Side by side) 21.1.6352462 (revision: 21.1.6352462)" complete.
"Install NDK (Side by side) 21.1.6352462 (revision: 21.1.6352462)" finished.
```

Since we're already checking the `sdkmanager` and the `licenses`, I don't believe adding an extra NDK check is necessary.

@dnfield WDYT? We settle on NDK version `21.1.6352462` which is consistent with the Android Gradle `4.1.0`? Or am I missing something related to NDK versioning here?

Reply:
> e.g., something like a gradle message about not finding CMake

CMake also gets automatically installed:

```
License for package CMake 3.10.2.4988404 accepted.
Preparing "Install CMake 3.10.2.4988404 (revision: 3.10.2)".
"Install CMake 3.10.2.4988404 (revision: 3.10.2)" ready.
Installing CMake 3.10.2.4988404 in /Users/dacoharkes/Library/Android/sdk/cmake/3.10.2.4988404
"Install CMake 3.10.2.4988404 (revision: 3.10.2)" complete.
"Install CMake 3.10.2.4988404 (revision: 3.10.2)" finished.
```

Projects that don't set a specific CMake version [are built with CMake 3.10.2](https://developer.android.com/studio/projects/install-ndk#vanilla_cmake).

In Flutter Desktop Linux/Windows we use CMake 3.10 as well.

We should strongly discourage anyone from using a newer version than 3.10.

Reply:
Somewhere we have a matrix of supported agp and Gradle versions with flutter.  Maybe @blasten knows. We should probably document that it will use the version of the ndk tied to agp. This still could cause problems for plugins that will want specific versions of the ndk but I'm not sure how much better we could really do there.

Reply:
This seems reasonable to me, but I would like to have LGTM from @blasten on this as well, especially if there's some doc we should update to indicate the NDK as part of a matrix of supported versions. I can't find the doc I was thinking of, maybe it wasn't needed anymore and deleted.

Reply:
> After running `flutter doctor` being happy with the Android toolchain and `flutter doctor --android-licenses` without an NDK installed, running an app using an FFI plugin will automatically download the NDK:

Just to make things sure, Do we need to manually run `flutter doctor --android-licenses` in order to make NDK automatically installed? I couldn't find relevent info in the docs. I looked in here: https://docs.flutter.dev/packages-and-plugins/developing-packages

Reply:
Anyone building for Android needs to accept the relevant licenses.

> I looked in here: https://docs.flutter.dev/packages-and-plugins/developing-packages

General build requirements for everyone using Flutter aren't documented on pages about developing packages, they are on pages about getting Flutter installed and set up.

Reply:
Thanks :) So from Flutter 3.0, when we develop an FFI plugin, it is guaranteed that NDK version required which is written in `./android/app/build.gradle` is installed **automatically**. It would be better to have this info on the docs IMHO.